File: src/Jhoose.Security.Core/Models/CSP/CspOptions.cs
using System.Text;

namespace Jhoose.Security.Core.Models.CSP
{
    public class CspOptions
    {
        public CspOptions()
        {
            this.None = false;
            this.Wildcard = false;
            this.Self = false;

            this.UnsafeEval = false;
            this.WasmUnsafeEval = false;
            this.UnsafeHashes = false;
            this.UnsafeInline = false;
            this.StrictDynamic = false;
            this.Nonce = false;
        }

        public bool None { get; set; }
        public bool Wildcard { get; set; }
        public bool Self { get; set; }
        public bool UnsafeEval { get; set; }
        public bool WasmUnsafeEval { get; set; }
        public bool UnsafeHashes { get; set; }
        public bool UnsafeInline { get; set; }
        public bool StrictDynamic { get; set; }
        public bool Nonce { get; set; }

        public override string ToString()
        {
            var sb = new StringBuilder();

            if (this.None)
            {
                sb.Append("'none' ");
                return sb.ToString();
            }
            else
            {
                if (this.Wildcard) sb.Append("* ");
                if (this.Self) sb.Append("'self' ");

                if (this.UnsafeEval) sb.Append("'unsafe-eval' ");
                if (this.WasmUnsafeEval) sb.Append("'wasm-unsafe-eval' ");
                if (this.UnsafeHashes) sb.Append("'unsafe-hashes' ");
                if (this.UnsafeInline) sb.Append("'unsafe-inline' ");
                if (this.StrictDynamic) sb.Append("'strict-dynamic' ");

                if (this.Nonce) sb.Append("'nonce-{0}' ");
            }

            return sb.ToString();
        }

        public bool HasOptions => this.None | this.Wildcard | this.Self | this.WasmUnsafeEval| this.UnsafeEval | this.UnsafeHashes | this.UnsafeInline | this.StrictDynamic | this.Nonce;
    }
}

File: src/Jhoose.Security/src/components/csp/CspOptions.js
import React, { useState } from 'react';
import { Checkbox, GridCell, GridRow } from "@episerver/ui-framework";

export function CspOptions(props) {

    var [options, setOptions] = useState(props.options);

    function setPolicyValue(key, value) {

        var newOptions = { ...options };

        newOptions[key] = !options[key];

        setOptions(newOptions);
        props.update(newOptions);
    };

    return (<fieldset>
        <legend>Options</legend>
        <GridRow>
            <GridCell span={12}>
                <Checkbox checked={options.none} onChange={(e) => {
                    setPolicyValue("none", e.currentTarget.value);
                }}>None</Checkbox>
            </GridCell>
            {!options.none &&
                <>
                    <GridCell span={3}>
                        <Checkbox checked={options.wildcard} onChange={(e) => {
                            setPolicyValue("wildcard", e.currentTarget.value);
                        }}>Wildcard</Checkbox>
                    </GridCell>
                    <GridCell span={3}>
                        <Checkbox checked={options.self} onChange={(e) => {
                            setPolicyValue("self", e.currentTarget.value);
                        }}>Self</Checkbox>
                    </GridCell>
                    {props.showScriptOptions &&
                    <>
                    <GridCell span={3}>
                        <Checkbox checked={options.nonce} onChange={(e) => {
                            setPolicyValue("nonce", e.currentTarget.value);
                        }}>Nonce</Checkbox>
                    </GridCell>
                    <GridCell span={3}>
                        <Checkbox checked={options.unsafeEval} onChange={(e) => {
                            setPolicyValue("unsafeEval", e.currentTarget.value);
                        }}>Unsafe Eval</Checkbox>
                    </GridCell>
                    <GridCell span={3}>
                        <Checkbox disabled={options.none} checked={options.unsafeHashes} onChange={(e) => {
                            setPolicyValue("unsafeHashes", e.currentTarget.value);
                        }}>Unsafe Hashes</Checkbox>
                    </GridCell>
                    <GridCell span={3}>
                        <Checkbox checked={options.unsafeInline} onChange={(e) => {
                            setPolicyValue("unsafeInline", e.currentTarget.value);
                        }}>Unsafe Inline</Checkbox>
                    </GridCell>
                    <GridCell span={3}>
                        <Checkbox checked={options.strictDynamic} onChange={(e) => {
                            setPolicyValue("strictDynamic", e.currentTarget.value);
                        }}>Strict Dynamic</Checkbox>
                    </GridCell>
                    </>}
                </>}
        </GridRow>

    </fieldset>);
}
